### PR TITLE
De-obfuscate Python

### PR DESCRIPTION
**Pull request by danpalmer**
Aims:
- Simplify control flow, be more Pythonic, take advantage of Python syntax, rather than implementing features that require Haskell syntax to be readable.
- Use built-in Python functionality instead of Haskell-style utils.
- More debuggable (better stacktraces).
- ...without sacrificing the speed too much.

This is very much a work in progress.

I realise the value of concepts like currying and composition, and in languages where they are first-class features (like Haskell), they work well because APIs are all designed with them in mind, and developers read the code assuming those concepts. However when applied to Python like this, they only serve to surprise, and obfuscate the true meaning of the code, making ongoing maintenance hard, and making it more difficult for new developers to contribute to the project. For this reason, I'm attempting to refactor dockhand into a more Pythonic form.

### CODE REVIEW

File: shipwright/build.py [hunk 1]
```diff
@@ -66,20 +67,26 @@ def process_event_(evt):
...
-
     >>> stream = iter(('Blah', 'Successfully built 1234\\n'))
     >>> success_from_stream(stream)
     '1234'
     """
+    for x in stream:
+        result = success(x)
+        if result:
+            return result
+    return None
```
Comment (graingert) [inline shipwright/build.py:114]: Unnecessary return

Comment (danpalmer) [inline shipwright/build.py:114]: I believe the API here is that the `None` is meaningful, so I think it's important to return it explicitly.

**Comment by danpalmer** [comment]
One thing that would help a lot with this is a wide range of unit and integration tests. Obviously tests against functions in `fn.py` would be a little useless, but it really would speed up this refactoring.

**Comment by graingert** [comment]
Can you squash/rebase to trigger a Travis build

File: shipwright/build.py [hunk 2]
```diff
@@ -50,9 +51,27 @@ def build(client, git_rev, container):
         'rev': git_rev
     }
 
-    def process_event_(evt):
-        evt_parsed = json_loads(evt)
-        return merge(merge_config)(evt_parsed)
+    # docker-py has an issue where it doesn't handle chunked responses from
```
Comment (graingert) [inline shipwright/build.py:54]: can you make this change in another PR

File: shipwright/build.py [hunk 3]
```diff
@@ -50,9 +51,27 @@ def build(client, git_rev, container):
...
+    # valid JSON objects that it is documented as yielding.
+    # We therefore maintain a buffer and read our own valid JSON out of that.
+    # This can probably be replaced this issue is fixed - ideally we'd use the
+    # 'decode' option on client.build to receive already parsed JSON objects.
+    #   https://github.com/docker/docker-py/issues/1059
+    buffer = b''
+
+    def process_event_(buffer, data):
+        buffer += data
+        for line in buffer.split(b'\r\n'):
```
Comment (graingert) [inline shipwright/build.py:65]: I'd go for the JSON lines approach http://jsonlines.org/

File: shipwright/build.py (hunk 2 above)
Comment (danpalmer) [inline shipwright/build.py:54]: I can, but it would need to be merged for me to be able to carry on working on this PR as its a build failure for me.

File: shipwright/build.py [hunk 4]
```diff
@@ -63,23 +82,33 @@ def process_event_(evt):
         dockerfile=os.path.basename(container.path),
     )
 
-    return (process_event_(evt) for evt in build_evts)
+    return (
```
Comment (graingert) [inline shipwright/build.py:85]: I'd probably go for nested for with yields.

File: shipwright/build.py (hunk 3 above)
Comment (danpalmer) [inline shipwright/build.py:65]: In what sense? This is informally JSON lines, and Docker isn't specifically using JSON lines.

Comment (graingert) [inline shipwright/build.py:65]: This looks like it might break on a Linux machine

File: shipwright/commits.py [hunk 5]
```diff
@@ -38,7 +38,7 @@ def key(item):
                 compat.python2_sort_key(c_map),
                 compat.python2_sort_key(ident),
             )
-        return max(([identity(c), commit_map(c)] for c in commits), key=key)
```
Comment (graingert) [inline shipwright/commits.py:41]: seriously

File: shipwright/build.py (hunk 3 above)
Comment (danpalmer) [inline shipwright/build.py:65]: Ok, so a `split('\n')` and a `strip`? I think the CRLF is from Docker regardless of the platform, but I don't know for sure - on my Mac it shouldn't be using CRs but it is.

File: shipwright/commits.py (hunk 5 above)
Comment (danpalmer) [inline shipwright/commits.py:41]: Yes, because `identity(c)` typechecks! ಠ_ಠ

File: shipwright/build.py (hunk 2 above)
Comment (graingert) [inline shipwright/build.py:54]: yeah makes sense.

**Comment by graingert** [comment]
mostly merged, I've extracted the Docker 1.11 fix https://github.com/graingert/dockhand/pull/4
